### PR TITLE
JPERF-273: Jira provisioning hooks

### PR DESCRIPTION
**Pull request by dagguh**
This is the 2023 version of #103
Rebased, resolved a lot of conflicts and cleaned up a little. Turns out many changes were done independently on `master` already. Still gotta squash them commits...

### CODE REVIEW

**Comment by dagguh** [comment]
Currently the `JiraServerPlanIT` fails with this: 
image
Maybe MySQL 5.7 is too new for this old Jira 7.13.0? Perhaps it's time to bump the default JSW version.

File: build.gradle.kts [hunk 1]
```diff
@@ -33,10 +33,16 @@ configurations.all {
...
                 "com.google.code.findbugs:jsr305" -> useVersion("3.0.2")
                 "org.apache.commons:commons-compress" -> useVersion("1.9")
+                "commons-io:commons-io" -> useVersion("2.6")
+                "org.apache.commons:commons-lang3" -> useVersion("3.12.0")
             }
             when (requested.group) {
                 "org.jetbrains.kotlin" -> useVersion(kotlinVersion)
                 "org.apache.logging.log4j" -> useVersion(log4jVersion)
+                "org.bouncycastle" -> useVersion("1.64")
+                // Gradle 5 treats this module as conflicted, even though the range is compatible
```
Comment (pczuj) [inline build.gradle.kts:43]: "this" is unclear is it about `bouncycastle` or `docker-java`

Comment (dagguh) [reply build.gradle.kts:43]:
```
// given
val x = 1
// when
sut.test(x)
```

Does `when` refer to `x = 1` or `sut.test`?
Line comments usually precede the code.

**Comment by dagguh** [comment]
Locally failing like this: https://scans.gradle.com/s/375dwevxnzzse

File: build.gradle.kts (hunk 1 above)
Comment (pczuj) [reply build.gradle.kts:43]: Fix would be as simple as changing "this" to `docker-java`

File: src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/jira/install/hook/PostInstallHooks.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    internal fun call(
+        ssh: SshConnection,
+        jira: InstalledJira,
+        reports: Reports
+    ) {
+        while (true) {
+            hooks
+                .poll()
+                ?.call(ssh, jira, this, reports)
+                ?: break
```
Comment (pczuj) [inline src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/jira/install/hook/PostInstallHooks.kt:35]: Nit: If I'm not wrong it's possible to create a PostInstallHook that'd create an infinite loop here.

File: src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/jira/instance/JiraServerPlan.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    private class JiraServer(
+        node: StartedJira
+    ) : JiraInstance {
+        override val address: URI = node.installed.http.addressPublicly()
+        override val nodes: List<StartedJira> = listOf(node)
+    }
+
+    class Builder(
+        private var plan: JiraNodePlan
+
```
Comment (pczuj) [inline src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/jira/instance/JiraServerPlan.kt:37]: nit: empty line

File: src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/jira/instance/PreInstanceHooks.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    private val hooks: Queue<PreInstanceHook> = ConcurrentLinkedQueue()
+
+    fun insert(hook: PreInstanceHook) {
+        hooks.add(hook)
+    }
+
+    internal fun call(nodes: List<PreInstallHooks>, reports: Reports) {
+        hooks
+            .parallelStream()
```
Comment (pczuj) [inline src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/jira/instance/PreInstanceHooks.kt:20]: Another usage of parallel stream. Isn't this risky?

File: src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/jira/sharedhome/NfsSharedHome.kt [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            download(ssh)
+            export(ssh)
+        }
+        nodes.forEach { it.postInstall.insert(NfsMount(tcp, localHome)) }
+    }
+
+    private fun download(ssh: SshConnection) {
+        ssh.execute("sudo mkdir -p $localHome")
+        val jiraHome = jiraHomeSource.download(ssh)
+        ssh.execute("sudo mv $jiraHome/{data,plugins,import,export} $localHome")
```
[the remark this answers is not in the document]
Comment (pczuj) [reply src/main/kotlin/com/atlassian/performance/tools/infrastructure/api/jira/sharedhome/NfsSharedHome.kt:37]: Fixing it in old API: https://github.com/atlassian/infrastructure/pull/149

Possibly same change should be applied here.

File: src/test/kotlin/com/atlassian/performance/tools/infrastructure/api/jira/instance/JiraPlanIT.kt [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+import org.assertj.core.api.SoftAssertions.assertSoftly
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import java.nio.file.Files
+import java.nio.file.Paths
+import java.time.Duration
+import java.time.Instant
+
+class JiraPlanIT {
```
Comment (pczuj) [inline src/test/kotlin/com/atlassian/performance/tools/infrastructure/api/jira/instance/JiraPlanIT.kt:31]: Would be good to have also a test for starting empty Jira (no dataset) without setup.
